File: app/src/main/java/com/soundcloud/android/discovery/DiscoveryCard.java
package com.soundcloud.android.discovery;

import com.google.auto.value.AutoValue;
import com.soundcloud.android.model.Urn;
import com.soundcloud.java.optional.Optional;

import java.util.List;

abstract class DiscoveryCard {
    enum Kind {
        SEARCH_ITEM,
        MULTIPLE_CONTENT_SELECTION_CARD,
        SINGLE_CONTENT_SELECTION_CARD,
        EMPTY_CARD
    }

    abstract Kind kind();

    static DiscoveryCard forSearchItem() {
        return DiscoveryCard.Default.create(Kind.SEARCH_ITEM);
    }

    boolean hasSelectionUrn(Urn selectionUrn) {
        return false;
    }

    @AutoValue
    abstract static class Default extends DiscoveryCard {
        public static DiscoveryCard create(Kind kind) {
            return new AutoValue_DiscoveryCard_Default(kind);
        }
    }

    @AutoValue
    abstract static class MultipleContentSelectionCard extends DiscoveryCard {

        abstract Urn selectionUrn();

        abstract Optional<Urn> queryUrn();

        abstract Optional<String> style();

        abstract Optional<String> title();

        abstract Optional<String> description();

        abstract List<SelectionItem> selectionItems();

        static MultipleContentSelectionCard create(Urn selectionUrn,
                                                   Optional<Urn> queryUrn,
                                                   Optional<String> style,
                                                   Optional<String> title,
                                                   Optional<String> description,
                                                   List<SelectionItem> selectionItems) {
            return new AutoValue_DiscoveryCard_MultipleContentSelectionCard(Kind.MULTIPLE_CONTENT_SELECTION_CARD, selectionUrn, queryUrn, style, title, description, selectionItems);
        }

        @Override
        boolean hasSelectionUrn(Urn selectionUrn) {
            return selectionUrn().equals(selectionUrn);
        }
    }

    @AutoValue
    abstract static class SingleContentSelectionCard extends DiscoveryCard {

        abstract Urn selectionUrn();

        abstract Optional<String> style();

        abstract Optional<String> title();

        abstract Optional<String> description();

        abstract Optional<Urn> queryUrn();

        abstract SelectionItem selectionItem();

        abstract Optional<String> socialProof();

        abstract List<String> socialProofAvatarUrlTemplates();

        static SingleContentSelectionCard create(Urn selectionUrn,
                                                 Optional<Urn> queryUrn,
                                                 Optional<String> style,
                                                 Optional<String> title,
                                                 Optional<String> description,
                                                 SelectionItem selectionItem,
                                                 Optional<String> socialProof,
                                                 List<String> socialProofAvatarUrlTemplates) {
            return new AutoValue_DiscoveryCard_SingleContentSelectionCard(Kind.SINGLE_CONTENT_SELECTION_CARD,
                    selectionUrn,
                    style,
                    title,
                    description,
                    queryUrn,
                    selectionItem,
                    socialProof,
                    socialProofAvatarUrlTemplates);
        }

        @Override
        boolean hasSelectionUrn(Urn selectionUrn) {
            return selectionUrn().equals(selectionUrn);
        }
    }

    @AutoValue
    abstract static class EmptyCard extends DiscoveryCard {

        abstract Optional<Throwable> throwable();

        static EmptyCard create(Optional<Throwable> throwable) {
            return new AutoValue_DiscoveryCard_EmptyCard(Kind.EMPTY_CARD, throwable);
        }
    }
}

File: app/src/main/java/com/soundcloud/android/discovery/DiscoveryCard.kt
package com.soundcloud.android.discovery

import com.soundcloud.android.model.Urn
import com.soundcloud.java.optional.Optional

sealed class DiscoveryCard {

    abstract val parentQueryUrn: Urn?

    internal open fun hasSelectionUrn(selectionUrn: Urn): Boolean = false

    data class MultipleContentSelectionCard(override val parentQueryUrn: Urn? = null,
                                            val selectionUrn: Urn,
                                            val queryUrn: Urn? = null,
                                            val style: String? = null,
                                            val title: String? = null,
                                            val description: String? = null,
                                            val trackingFeatureName: String? = null,
                                            val selectionItems: List<SelectionItem>) : DiscoveryCard() {

        override fun hasSelectionUrn(selectionUrn: Urn): Boolean = this.selectionUrn == selectionUrn
    }

    data class SingleContentSelectionCard(override val parentQueryUrn: Urn? = null,
                                          val selectionUrn: Urn,
                                          val style: String? = null,
                                          val title: String? = null,
                                          val description: String? = null,
                                          val queryUrn: Urn? = null,
                                          val selectionItem: SelectionItem,
                                          val trackingFeatureName: String? = null,
                                          val socialProof: String? = null,
                                          val socialProofAvatarUrlTemplates: List<String>) : DiscoveryCard() {

        override fun hasSelectionUrn(selectionUrn: Urn): Boolean = this.selectionUrn == selectionUrn
    }

    data class EmptyCard(val throwable: Throwable? = null, override val parentQueryUrn: Urn? = null) : DiscoveryCard() {
        constructor(optionalThrowable: Optional<Throwable>) : this(optionalThrowable.orNull())
    }
}
